patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch

```diff
@@ -17,23 +17,189 @@
+             this.field_70963_i = new MerchantRecipeList(nbttagcompound1);
+         }
+@@ -333,14 +356,27 @@
  
-     private void func_175554_cu()
+     public void func_70938_b(int p_70938_1_)
+     {
++        this.field_175563_bv = 0;
++        this.field_175562_bw = 0;
++        this.field_70963_i = null;
+         this.field_70180_af.func_75692_b(16, Integer.valueOf(p_70938_1_));
+     }
+ 
+     public int func_70946_n()
+     {
+-        return Math.max(this.field_70180_af.func_75679_c(16) % 5, 0);
++        return Math.max(this.field_70180_af.func_75679_c(16) % net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession.nextId, 0);
+     }
+ 
++    public void setCareer(int careerId)
++    {
++        this.field_175563_bv = careerId;
++    }
++
++    public int getCareer()
++    {
++        return this.field_175563_bv;
++    }
++
+     public boolean func_70941_o()
      {
-+        //TODO: Hook into VillagerRegistry
-         EntityVillager.ITradeList[][][] aitradelist = field_175561_bA[this.func_70946_n()];
+         return this.field_70952_f;
+@@ -540,15 +576,14 @@
  
+     private void func_175554_cu()
+     {
+-        EntityVillager.ITradeList[][][] aitradelist = field_175561_bA[this.func_70946_n()];
+-
          if (this.field_175563_bv != 0 && this.field_175562_bw != 0)
-@@ -648,6 +650,7 @@
-                     }
+         {
+             ++this.field_175562_bw;
+         }
+         else
+         {
+-            this.field_175563_bv = this.field_70146_Z.nextInt(aitradelist.length) + 1;
++            int careerCount = net.minecraftforge.fml.common.registry.VillagerRegistry.getNumberOfCareers(this.func_70946_n());
++            this.field_175563_bv = this.field_70146_Z.nextInt(careerCount) + 1;
+             this.field_175562_bw = 1;
+         }
+ 
+@@ -556,23 +591,7 @@
+         {
+             this.field_70963_i = new MerchantRecipeList();
+         }
+-
+-        int i = this.field_175563_bv - 1;
+-        int j = this.field_175562_bw - 1;
+-        EntityVillager.ITradeList[][] aitradelist1 = aitradelist[i];
+-
+-        if (j < aitradelist1.length)
+-        {
+-            EntityVillager.ITradeList[] aitradelist2 = aitradelist1[j];
+-            EntityVillager.ITradeList[] aitradelist3 = aitradelist2;
+-            int k = aitradelist2.length;
+-
+-            for (int l = 0; l < k; ++l)
+-            {
+-                EntityVillager.ITradeList itradelist = aitradelist3[l];
+-                itradelist.func_179401_a(this.field_70963_i, this.field_70146_Z);
+-            }
+-        }
++        net.minecraftforge.fml.common.registry.VillagerRegistry.populateBuyingList(this.field_70963_i, this.func_70946_n(), this.field_175563_bv, this.field_175562_bw, this.field_70146_Z);
+     }
+ 
+     @SideOnly(Side.CLIENT)
+@@ -593,61 +612,7 @@
+                 this.func_175554_cu();
              }
  
-+            //TODO: Hook into VillagerRegistry to get name
+-            String s1 = null;
+-
```

Comment:
This patch segment needs to be a lot smaller, no need to remove the switch if you just override it down below.